Unofficial package doesn't work in Visual Studio 2015 Update 1

```
error MSB4018: The "TransformOnBuildTask" task failed unexpectedly.
error MSB4018: System.Exception: Could not extract targetBuilderCallback from Microsoft.Build.BackEnd.TaskHost
error MSB4018:    at Clarius.TransformOnBuild.MSBuild.Task.TransformOnBuildTask.GetProjectInstance()
error MSB4018:    at Clarius.TransformOnBuild.MSBuild.Task.TransformOnBuildTask.Execute()
error MSB4018:    at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute()
error MSB4018:    at Microsoft.Build.BackEnd.TaskBuilder.<ExecuteInstantiatedTask>d__26.MoveNext()
```

--- a/.gitignore
+++ b/.gitignore
@@ -1,2 +1,30 @@
 /.nuget
 *.nupkg
+#ignore thumbnails created by windows
+Thumbs.db
+#Ignore files build by Visual Studio
+*.obj
+*.exe
+*.pdb
+*.user
+*.aps
+*.pch
+*.vspscc
+*_i.c
+*_p.c
+*.ncb
+*.suo
+*.tlb
+*.tlh
+*.bak
+*.cache
+*.ilk
+*.log
+[Bb]in
+[Dd]ebug*/
+*.lib
+*.sbr
+obj/
+[Rr]elease*/
+_ReSharper*/
+[Tt]est[Rr]esult*

--- a/.gitignore
+++ b/.gitignore
@@ -28,3 +28,4 @@ obj/
 [Rr]elease*/
 _ReSharper*/
 [Tt]est[Rr]esult*
+nuget/build/Clarius.TransformOnBuild.MSBuild.Task.dll

--- a/Clarius.TransformOnBuild.MSBuild.Task/Clarius.TransformOnBuild.MSBuild.Task.sln
+++ b/Clarius.TransformOnBuild.MSBuild.Task/Clarius.TransformOnBuild.MSBuild.Task.sln
@@ -1,0 +1,22 @@
+﻿
+Microsoft Visual Studio Solution File, Format Version 12.00
+# Visual Studio 2013
+VisualStudioVersion = 12.0.31101.0
+MinimumVisualStudioVersion = 10.0.40219.1
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Clarius.TransformOnBuild.MSBuild.Task", "Clarius.TransformOnBuild.MSBuild.Task\Clarius.TransformOnBuild.MSBuild.Task.csproj", "{EB0AFC95-520D-4F4A-9709-E1EA69704954}"
+EndProject
+Global
+	GlobalSection(SolutionConfigurationPlatforms) = preSolution
+		Debug|Any CPU = Debug|Any CPU
+		Release|Any CPU = Release|Any CPU
+	EndGlobalSection
+	GlobalSection(ProjectConfigurationPlatforms) = postSolution
+		{EB0AFC95-520D-4F4A-9709-E1EA69704954}.Debug|Any CPU.ActiveCfg = Debug|Any CPU
+		{EB0AFC95-520D-4F4A-9709-E1EA69704954}.Debug|Any CPU.Build.0 = Debug|Any CPU
+		{EB0AFC95-520D-4F4A-9709-E1EA69704954}.Release|Any CPU.ActiveCfg = Release|Any CPU
+		{EB0AFC95-520D-4F4A-9709-E1EA69704954}.Release|Any CPU.Build.0 = Release|Any CPU
+	EndGlobalSection
+	GlobalSection(SolutionProperties) = preSolution
+		HideSolutionNode = FALSE
+	EndGlobalSection
+EndGlobal

--- a/Clarius.TransformOnBuild.MSBuild.Task/Clarius.TransformOnBuild.MSBuild.Task/Clarius.TransformOnBuild.MSBuild.Task.csproj
+++ b/Clarius.TransformOnBuild.MSBuild.Task/Clarius.TransformOnBuild.MSBuild.Task/Clarius.TransformOnBuild.MSBuild.Task.csproj
@@ -13,8 +13,8 @@
     <FileAlignment>512</FileAlignment>
   </PropertyGroup>
   <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Debug|AnyCPU' ">
-    <DebugSymbols>true</DebugSymbols>
-    <DebugType>full</DebugType>
+    <DebugSymbols>false</DebugSymbols>
+    <DebugType>none</DebugType>
     <Optimize>false</Optimize>
     <OutputPath>..\..\nuget\build\</OutputPath>
     <DefineConstants>DEBUG;TRACE</DefineConstants>

--- a/Clarius.TransformOnBuild.MSBuild.Task/Clarius.TransformOnBuild.MSBuild.Task/Clarius.TransformOnBuild.MSBuild.Task.csproj
+++ b/Clarius.TransformOnBuild.MSBuild.Task/Clarius.TransformOnBuild.MSBuild.Task/Clarius.TransformOnBuild.MSBuild.Task.csproj
@@ -1,0 +1,56 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<Project ToolsVersion="12.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <Import Project="$(MSBuildExtensionsPath)\$(MSBuildToolsVersion)\Microsoft.Common.props" Condition="Exists('$(MSBuildExtensionsPath)\$(MSBuildToolsVersion)\Microsoft.Common.props')" />
+  <PropertyGroup>
+    <Configuration Condition=" '$(Configuration)' == '' ">Debug</Configuration>
+    <Platform Condition=" '$(Platform)' == '' ">AnyCPU</Platform>
+    <ProjectGuid>{EB0AFC95-520D-4F4A-9709-E1EA69704954}</ProjectGuid>
+    <OutputType>Library</OutputType>
+    <AppDesignerFolder>Properties</AppDesignerFolder>
+    <RootNamespace>Clarius.TransformOnBuild.MSBuild.Task</RootNamespace>
+    <AssemblyName>Clarius.TransformOnBuild.MSBuild.Task</AssemblyName>
+    <TargetFrameworkVersion>v4.0</TargetFrameworkVersion>
+    <FileAlignment>512</FileAlignment>
+  </PropertyGroup>
+  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Debug|AnyCPU' ">
+    <DebugSymbols>true</DebugSymbols>
+    <DebugType>full</DebugType>
+    <Optimize>false</Optimize>
+    <OutputPath>..\..\nuget\build\</OutputPath>
+    <DefineConstants>DEBUG;TRACE</DefineConstants>
+    <ErrorReport>prompt</ErrorReport>
+    <WarningLevel>4</WarningLevel>
+  </PropertyGroup>
+  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Release|AnyCPU' ">
+    <DebugType>pdbonly</DebugType>
+    <Optimize>true</Optimize>
+    <OutputPath>bin\Release\</OutputPath>
+    <DefineConstants>TRACE</DefineConstants>
+    <ErrorReport>prompt</ErrorReport>
+    <WarningLevel>4</WarningLevel>
+  </PropertyGroup>
+  <ItemGroup>
+    <Reference Include="Microsoft.Build" />
+    <Reference Include="Microsoft.Build.Framework" />
+    <Reference Include="Microsoft.Build.Utilities.v4.0" />
+    <Reference Include="System" />
+    <Reference Include="System.Core" />
+    <Reference Include="System.Xml.Linq" />
+    <Reference Include="System.Data.DataSetExtensions" />
+    <Reference Include="Microsoft.CSharp" />
+    <Reference Include="System.Data" />
+    <Reference Include="System.Xml" />
+  </ItemGroup>
+  <ItemGroup>
+    <Compile Include="TransformOnBuildTask.cs" />
+    <Compile Include="Properties\AssemblyInfo.cs" />
+  </ItemGroup>
+  <Import Project="$(MSBuildToolsPath)\Microsoft.CSharp.targets" />
+  <!-- To modify your build process, add your task inside one of the targets below and uncomment it. 
+       Other similar extension points exist, see Microsoft.Common.targets.
+  <Target Name="BeforeBuild">
+  </Target>
+  <Target Name="AfterBuild">
+  </Target>
+  -->
+</Project>

--- a/Clarius.TransformOnBuild.MSBuild.Task/Clarius.TransformOnBuild.MSBuild.Task/Properties/AssemblyInfo.cs
+++ b/Clarius.TransformOnBuild.MSBuild.Task/Clarius.TransformOnBuild.MSBuild.Task/Properties/AssemblyInfo.cs
@@ -1,0 +1,36 @@
+﻿using System.Reflection;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+// General Information about an assembly is controlled through the following 
+// set of attributes. Change these attribute values to modify the information
+// associated with an assembly.
+[assembly: AssemblyTitle("Clarius.TransformOnBuild.MSBuild.Task")]
+[assembly: AssemblyDescription("")]
+[assembly: AssemblyConfiguration("")]
+[assembly: AssemblyCompany("")]
+[assembly: AssemblyProduct("Clarius.TransformOnBuild.MSBuild.Task")]
+[assembly: AssemblyCopyright("Copyright ©  2015")]
+[assembly: AssemblyTrademark("")]
+[assembly: AssemblyCulture("")]
+
+// Setting ComVisible to false makes the types in this assembly not visible 
+// to COM components.  If you need to access a type in this assembly from 
+// COM, set the ComVisible attribute to true on that type.
+[assembly: ComVisible(false)]
+
+// The following GUID is for the ID of the typelib if this project is exposed to COM
+[assembly: Guid("222c7308-d35f-49a3-bedc-ea351ac8dbf9")]
+
+// Version information for an assembly consists of the following four values:
+//
+//      Major Version
+//      Minor Version 
+//      Build Number
+//      Revision
+//
+// You can specify all the values or you can default the Build and Revision Numbers 
+// by using the '*' as shown below:
+// [assembly: AssemblyVersion("1.0.*")]
+[assembly: AssemblyVersion("1.0.0.0")]
+[assembly: AssemblyFileVersion("1.0.0.0")]

--- a/Clarius.TransformOnBuild.MSBuild.Task/Clarius.TransformOnBuild.MSBuild.Task/TransformOnBuildTask.cs
+++ b/Clarius.TransformOnBuild.MSBuild.Task/Clarius.TransformOnBuild.MSBuild.Task/TransformOnBuildTask.cs
@@ -163,12 +163,12 @@ namespace Clarius.TransformOnBuild.MSBuild.Task
         private ProjectInstance GetProjectInstance()
         {
             var buildEngineType = BuildEngine.GetType();
-            var targetBuilderCallbackField = buildEngineType.GetField("targetBuilderCallback", BindingFlags);
+            var targetBuilderCallbackField = buildEngineType.GetField("targetBuilderCallback", BindingFlags) ?? buildEngineType.GetField("_targetBuilderCallback", BindingFlags);
             if (targetBuilderCallbackField == null)
                 throw new Exception("Could not extract targetBuilderCallback from " + buildEngineType.FullName);
             var targetBuilderCallback = targetBuilderCallbackField.GetValue(BuildEngine);
             var targetCallbackType = targetBuilderCallback.GetType();
-            var projectInstanceField = targetCallbackType.GetField("projectInstance", BindingFlags);
+            var projectInstanceField = targetCallbackType.GetField("projectInstance", BindingFlags) ?? targetCallbackType.GetField("_projectInstance", BindingFlags);
             if (projectInstanceField == null)
                 throw new Exception("Could not extract projectInstance from " + targetCallbackType.FullName);
             return (ProjectInstance) projectInstanceField.GetValue(targetBuilderCallback);

--- a/Clarius.TransformOnBuild.MSBuild.Task/Clarius.TransformOnBuild.MSBuild.Task/TransformOnBuildTask.cs
+++ b/Clarius.TransformOnBuild.MSBuild.Task/Clarius.TransformOnBuild.MSBuild.Task/TransformOnBuildTask.cs
@@ -73,7 +73,7 @@ namespace Clarius.TransformOnBuild.MSBuild.Task
 
         private string RewriteTemplateContent(string template)
         {
-            var result = Regex.Replace(template, @"(?im)(<#@\s*assembly\s+name\s+="".*?""|<#@\s*include\s+file\s+="".*?"")",
+            var result = Regex.Replace(template, @"(?im)(<#@\s*assembly\s+name\s*=\s*"".*?""|<#@\s*include\s+file\s*=\s*"".*?"")",
                 m => ExpandVariables(m.Value));
             return result;
         }

--- a/Clarius.TransformOnBuild.MSBuild.Task/Clarius.TransformOnBuild.MSBuild.Task/TransformOnBuildTask.cs
+++ b/Clarius.TransformOnBuild.MSBuild.Task/Clarius.TransformOnBuild.MSBuild.Task/TransformOnBuildTask.cs
@@ -38,5 +38,15 @@ namespace Clarius.TransformOnBuild.MSBuild.Task
                 throw new Exception("Could not extract projectInstance from " + targetCallbackType.FullName);
             return (ProjectInstance) projectInstanceField.GetValue(targetBuilderCallback);
         }
+
+        private string GetPropertyValue(string propertyName, bool throwIfNotFound = false)
+        {
+            string propertyValue;
+            if (_properties.TryGetValue(propertyName, out propertyValue))
+                return propertyValue;
+            if (throwIfNotFound)
+                throw new Exception(string.Format("Could not resolve property $({0})", propertyName));
+            return "";
+        }
     }
 }

--- a/Clarius.TransformOnBuild.MSBuild.Task/Clarius.TransformOnBuild.MSBuild.Task/TransformOnBuildTask.cs
+++ b/Clarius.TransformOnBuild.MSBuild.Task/Clarius.TransformOnBuild.MSBuild.Task/TransformOnBuildTask.cs
@@ -4,7 +4,6 @@ using System.Diagnostics;
 using System.IO;
 using System.Linq;
 using System.Reflection;
-using System.Text;
 using System.Text.RegularExpressions;
 using Microsoft.Build.Execution;
 using Microsoft.Build.Framework;

--- a/Clarius.TransformOnBuild.MSBuild.Task/Clarius.TransformOnBuild.MSBuild.Task/TransformOnBuildTask.cs
+++ b/Clarius.TransformOnBuild.MSBuild.Task/Clarius.TransformOnBuild.MSBuild.Task/TransformOnBuildTask.cs
@@ -36,7 +36,7 @@ namespace Clarius.TransformOnBuild.MSBuild.Task
             var textTransform = _projectInstance.Items.Where(i =>
             {
                 var comparer = StringComparer.InvariantCultureIgnoreCase;
-                return comparer.Equals(i.ItemType, "None") &&
+                return (comparer.Equals(i.ItemType, "None") || comparer.Equals(i.ItemType, "Content")) &&
                        comparer.Equals(i.GetMetadataValue("Generator"), "TextTemplatingFileGenerator");
             });
 

--- a/Clarius.TransformOnBuild.MSBuild.Task/Clarius.TransformOnBuild.MSBuild.Task/TransformOnBuildTask.cs
+++ b/Clarius.TransformOnBuild.MSBuild.Task/Clarius.TransformOnBuild.MSBuild.Task/TransformOnBuildTask.cs
@@ -41,11 +41,21 @@ namespace Clarius.TransformOnBuild.MSBuild.Task
             foreach (var templateItem in textTransform)
             {
                 var templatePath = templateItem.GetMetadataValue("FullPath");
+                var templateBackupPath = templatePath + ".bak_clarius";
+                try
+                {
+                    File.Copy(templatePath, templateBackupPath, overwrite: true);
 
-                var result = RunTransformTool(templatePath);
+                    var result = RunTransformTool(templatePath);
 
-                if (!result)
-                    return false;
+                    if (!result)
+                        return false;
+                }
+                finally
+                {
+                    File.Copy(templateBackupPath, templatePath, overwrite: true);
+                    File.Delete(templateBackupPath);
+                }
             }
 
             return true;

--- a/Clarius.TransformOnBuild.MSBuild.Task/Clarius.TransformOnBuild.MSBuild.Task/TransformOnBuildTask.cs
+++ b/Clarius.TransformOnBuild.MSBuild.Task/Clarius.TransformOnBuild.MSBuild.Task/TransformOnBuildTask.cs
@@ -1,5 +1,6 @@
 ﻿using System;
 using System.Collections.Generic;
+using System.IO;
 using System.Linq;
 using System.Reflection;
 using Microsoft.Build.Execution;
@@ -10,6 +11,9 @@ namespace Clarius.TransformOnBuild.MSBuild.Task
     {
         private ProjectInstance _projectInstance;
         private Dictionary<string, string> _properties;
+        private string _commonProgramFiles;
+        private string _textTransformPath;
+        private string _transformExe;
         const BindingFlags BindingFlags = System.Reflection.BindingFlags.NonPublic | System.Reflection.BindingFlags.FlattenHierarchy | System.Reflection.BindingFlags.Instance | System.Reflection.BindingFlags.Public;
 
         public override bool Execute()
@@ -17,8 +21,36 @@ namespace Clarius.TransformOnBuild.MSBuild.Task
             _projectInstance = GetProjectInstance();
             _properties = _projectInstance.Properties.ToDictionary(p => p.Name, p => p.EvaluatedValue);
 
+            InitPathProperties();
 
             return true;
+        }
+
+        private void InitPathProperties()
+        {
+            _commonProgramFiles = Environment.GetEnvironmentVariable("CommonProgramFiles(x86)");
+            if (string.IsNullOrEmpty(_commonProgramFiles))
+                _commonProgramFiles = GetPropertyValue("CommonProgramFiles");
+
+            _textTransformPath = GetPropertyValue("TextTransformPath");
+            if (string.IsNullOrEmpty(_textTransformPath))
+                _textTransformPath = string.Format(@"{0}\Microsoft Shared\TextTemplating\{1}\TextTransform.exe", _commonProgramFiles, GetPropertyValue("VisualStudioVersion"));
+
+            // Initial default value
+            _transformExe = _textTransformPath;
+
+            // Cascading probing if file not found
+            if (!File.Exists(_transformExe))
+                _transformExe = string.Format(@"{0}\Microsoft Shared\TextTemplating\10.0\TextTransform.exe", _commonProgramFiles);
+            if (!File.Exists(_transformExe))
+                _transformExe = string.Format(@"{0}\Microsoft Shared\TextTemplating\11.0\TextTransform.exe", _commonProgramFiles);
+            if (!File.Exists(_transformExe))
+                _transformExe = string.Format(@"{0}\Microsoft Shared\TextTemplating\12.0\TextTransform.exe", _commonProgramFiles);
+            // Future proof 'til VS2013+2
+            if (!File.Exists(_transformExe))
+                _transformExe = string.Format(@"{0}\Microsoft Shared\TextTemplating\13.0\TextTransform.exe", _commonProgramFiles);
+            if (!File.Exists(_transformExe))
+                _transformExe = string.Format(@"{0}\Microsoft Shared\TextTemplating\14.0\TextTransform.exe", _commonProgramFiles);
         }
 
         /// <summary>

--- a/Clarius.TransformOnBuild.MSBuild.Task/Clarius.TransformOnBuild.MSBuild.Task/TransformOnBuildTask.cs
+++ b/Clarius.TransformOnBuild.MSBuild.Task/Clarius.TransformOnBuild.MSBuild.Task/TransformOnBuildTask.cs
@@ -1,0 +1,12 @@
+﻿using System;
+
+namespace Clarius.TransformOnBuild.MSBuild.Task
+{
+    public class TransformOnBuildTask : Microsoft.Build.Utilities.Task
+    {
+        public override bool Execute()
+        {
+            throw new NotImplementedException();
+        }
+    }
+}

--- a/Clarius.TransformOnBuild.MSBuild.Task/Clarius.TransformOnBuild.MSBuild.Task/TransformOnBuildTask.cs
+++ b/Clarius.TransformOnBuild.MSBuild.Task/Clarius.TransformOnBuild.MSBuild.Task/TransformOnBuildTask.cs
@@ -4,6 +4,7 @@ using System.Diagnostics;
 using System.IO;
 using System.Linq;
 using System.Reflection;
+using System.Text;
 using System.Text.RegularExpressions;
 using Microsoft.Build.Execution;
 using Microsoft.Build.Framework;
@@ -66,9 +67,19 @@ namespace Clarius.TransformOnBuild.MSBuild.Task
 
         private void RewriteTemplateFile(string templatePath)
         {
-            var template = File.ReadAllText(templatePath);
+            var encoding = GetCurrentEncoding(templatePath);
+            var template = File.ReadAllText(templatePath, encoding);
             template = RewriteTemplateContent(template);
-            File.WriteAllText(templatePath, template);
+            File.WriteAllText(templatePath, template, encoding);
+        }
+
+        private static Encoding GetCurrentEncoding(string templatePath)
+        {
+            using (var streamReader = new StreamReader(templatePath))
+            {
+                streamReader.ReadLine();
+                return streamReader.CurrentEncoding;
+            }
         }
 
         private string RewriteTemplateContent(string template)

--- a/Clarius.TransformOnBuild.MSBuild.Task/Clarius.TransformOnBuild.MSBuild.Task/TransformOnBuildTask.cs
+++ b/Clarius.TransformOnBuild.MSBuild.Task/Clarius.TransformOnBuild.MSBuild.Task/TransformOnBuildTask.cs
@@ -1,9 +1,11 @@
 ﻿using System;
 using System.Collections.Generic;
+using System.Diagnostics;
 using System.IO;
 using System.Linq;
 using System.Reflection;
 using Microsoft.Build.Execution;
+using Microsoft.Build.Framework;
 
 namespace Clarius.TransformOnBuild.MSBuild.Task
 {
@@ -23,7 +25,63 @@ namespace Clarius.TransformOnBuild.MSBuild.Task
 
             InitPathProperties();
 
+            if (!File.Exists(_transformExe))
+            {
+                Log.LogError("Failed to find TextTransform.exe tool at '{0}'.", _transformExe);
+                return false;
+            }
+
+            var textTransform = _projectInstance.Items.Where(i =>
+            {
+                var comparer = StringComparer.InvariantCultureIgnoreCase;
+                return comparer.Equals(i.ItemType, "None") &&
+                       comparer.Equals(i.GetMetadataValue("Generator"), "TextTemplatingFileGenerator");
+            });
+
+            foreach (var templateItem in textTransform)
+            {
+                var templatePath = templateItem.GetMetadataValue("FullPath");
+
+                var result = RunTransformTool(templatePath);
+
+                if (!result)
+                    return false;
+            }
+
             return true;
+        }
+
+        private bool RunTransformTool(string templatePath)
+        {
+            var process = new Process
+            {
+                StartInfo = new ProcessStartInfo
+                {
+                    FileName = _transformExe,
+                    Arguments = templatePath,
+                    RedirectStandardOutput = true,
+                    RedirectStandardError = true,
+                    CreateNoWindow = true,
+                    UseShellExecute = false
+                }
+            };
+            process.ErrorDataReceived += (sender, args) =>
+            {
+                if (args.Data == null)
+                    return;
+                Log.LogMessageFromText(args.Data, MessageImportance.Normal);
+            };
+            process.OutputDataReceived += (sender, args) =>
+            {
+                if (args.Data == null)
+                    return;
+                Log.LogMessageFromText(args.Data, MessageImportance.Low);
+            };
+            process.Start();
+            process.BeginErrorReadLine();
+            process.BeginOutputReadLine();
+            process.WaitForExit();
+            return process.ExitCode == 0;
         }
 
         private void InitPathProperties()

--- a/Clarius.TransformOnBuild.MSBuild.Task/Clarius.TransformOnBuild.MSBuild.Task/TransformOnBuildTask.cs
+++ b/Clarius.TransformOnBuild.MSBuild.Task/Clarius.TransformOnBuild.MSBuild.Task/TransformOnBuildTask.cs
@@ -1,12 +1,42 @@
 ﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using Microsoft.Build.Execution;
 
 namespace Clarius.TransformOnBuild.MSBuild.Task
 {
     public class TransformOnBuildTask : Microsoft.Build.Utilities.Task
     {
+        private ProjectInstance _projectInstance;
+        private Dictionary<string, string> _properties;
+        const BindingFlags BindingFlags = System.Reflection.BindingFlags.NonPublic | System.Reflection.BindingFlags.FlattenHierarchy | System.Reflection.BindingFlags.Instance | System.Reflection.BindingFlags.Public;
+
         public override bool Execute()
         {
-            throw new NotImplementedException();
+            _projectInstance = GetProjectInstance();
+            _properties = _projectInstance.Properties.ToDictionary(p => p.Name, p => p.EvaluatedValue);
+
+
+            return true;
+        }
+
+        /// <summary>
+        /// Inspired by http://stackoverflow.com/questions/3043531/when-implementing-a-microsoft-build-utilities-task-how-to-i-get-access-to-the-va
+        /// </summary>
+        /// <returns></returns>
+        private ProjectInstance GetProjectInstance()
+        {
+            var buildEngineType = BuildEngine.GetType();
+            var targetBuilderCallbackField = buildEngineType.GetField("targetBuilderCallback", BindingFlags);
+            if (targetBuilderCallbackField == null)
+                throw new Exception("Could not extract targetBuilderCallback from " + buildEngineType.FullName);
+            var targetBuilderCallback = targetBuilderCallbackField.GetValue(BuildEngine);
+            var targetCallbackType = targetBuilderCallback.GetType();
+            var projectInstanceField = targetCallbackType.GetField("projectInstance", BindingFlags);
+            if (projectInstanceField == null)
+                throw new Exception("Could not extract projectInstance from " + targetCallbackType.FullName);
+            return (ProjectInstance) projectInstanceField.GetValue(targetBuilderCallback);
         }
     }
 }

--- a/Clarius.TransformOnBuild.MSBuild.Task/Clarius.TransformOnBuild.MSBuild.Task/TransformOnBuildTask.cs
+++ b/Clarius.TransformOnBuild.MSBuild.Task/Clarius.TransformOnBuild.MSBuild.Task/TransformOnBuildTask.cs
@@ -4,6 +4,8 @@ using System.Diagnostics;
 using System.IO;
 using System.Linq;
 using System.Reflection;
+using System.Text;
+using System.Text.RegularExpressions;
 using Microsoft.Build.Execution;
 using Microsoft.Build.Framework;
 
@@ -46,6 +48,8 @@ namespace Clarius.TransformOnBuild.MSBuild.Task
                 {
                     File.Copy(templatePath, templateBackupPath, overwrite: true);
 
+                    RewriteTemplateFile(templatePath);
+
                     var result = RunTransformTool(templatePath);
 
                     if (!result)
@@ -59,6 +63,27 @@ namespace Clarius.TransformOnBuild.MSBuild.Task
             }
 
             return true;
+        }
+
+        private void RewriteTemplateFile(string templatePath)
+        {
+            var template = File.ReadAllText(templatePath);
+            template = RewriteTemplateContent(template);
+            File.WriteAllText(templatePath, template);
+        }
+
+        private string RewriteTemplateContent(string template)
+        {
+            var result = Regex.Replace(template, @"(?im)(<#@\s*assembly\s+name\s+="".*?""|<#@\s*include\s+file\s+="".*?"")",
+                m => ExpandVariables(m.Value));
+            return result;
+        }
+
+        private string ExpandVariables(string str)
+        {
+            var result = Environment.ExpandEnvironmentVariables(str);
+            result = Regex.Replace(result, @"\$\((?<PropertyName>.+?)\)", m => GetPropertyValue(m.Groups["PropertyName"].Value, throwIfNotFound: true));
+            return result;
         }
 
         private bool RunTransformTool(string templatePath)

--- a/README.md
+++ b/README.md
@@ -1,7 +1,7 @@
 ![Icon](https://raw.github.com/clariuslabs/TransformOnBuild/master/icon/32.png) Transform Text Templates On Build
 ============
 
-Automatically transforms on build all files with a build action of `None` that have the `TextTemplatingFileGenerator` custom tools associated.
+Automatically transforms on build all files with a build action of `None/Content` that have the `TextTemplatingFileGenerator` custom tools associated.
 
 ## Installation
 

--- a/build.proj
+++ b/build.proj
@@ -36,7 +36,15 @@
 		<PackageConfig Include="**\packages.config" />
 	</ItemGroup>
 		
-    <Target Name="Build" DependsOnTargets="BuildNuGet;PushNuGet" />
+    <Target Name="Build" DependsOnTargets="BuildMSBuildTask;BuildNuGet;PushNuGet" />
+
+    <Target
+      Name="BuildMSBuildTask">
+      <MSBuild
+        Projects="Clarius.TransformOnBuild.MSBuild.Task\Clarius.TransformOnBuild.MSBuild.Task.sln"
+      />
+    </Target>
+
 
 	<Target Name="BuildNuGet" Inputs="@(NuSpec)" Outputs="%(Identity)-BATCH" DependsOnTargets="_DownloadNuGet;RestorePackages">      
 		<PropertyGroup>

--- a/nuget/Clarius.TransformOnBuild.nuspec
+++ b/nuget/Clarius.TransformOnBuild.nuspec
@@ -4,7 +4,7 @@
 		<id>Clarius.TransformOnBuild</id>
 		<version>1.1</version>
 		<title>Transform Text Templates On Build</title>
-		<description>Automatically transforms on build all files with a build action of `None` that have the 
+		<description>Automatically transforms on build all files with a build action of `None/Content` that have the 
         `TextTemplatingFilePreprocessor` or `TextTemplatingFileGenerator` custom tools associated, without 
         requiring the installation of any Visual Studio SDK.</description>
 		<authors>Daniel Cazzulino</authors>

--- a/nuget/build/Clarius.TransformOnBuild.targets
+++ b/nuget/build/Clarius.TransformOnBuild.targets
@@ -24,6 +24,9 @@
   <Target
     Name="TransformOnBuild"
     AfterTargets="BeforeCompile">
+    <CallTarget
+      Targets="_CopyFilesMarkedCopyLocal"
+    />
     <TransformOnBuildTask />
   </Target>
 </Project>

--- a/nuget/build/Clarius.TransformOnBuild.targets
+++ b/nuget/build/Clarius.TransformOnBuild.targets
@@ -31,7 +31,7 @@
         <_TransformExe Condition="!Exists('$(_TransformExe)')">$(_CommonProgramFiles)\Microsoft Shared\TextTemplating\14.0\TextTransform.exe</_TransformExe>
 	</PropertyGroup>
 
-    <Target Name="TransformOnBuild" AfterTargets="BeforeBuild">
+    <Target Name="TransformOnBuild" AfterTargets="BeforeCompile">
 
         <Error Text="Failed to find TextTransform.exe tool at '$(_TransformExe)." 
                 Condition="!Exists('$(_TransformExe)')"/>

--- a/nuget/build/Clarius.TransformOnBuild.targets
+++ b/nuget/build/Clarius.TransformOnBuild.targets
@@ -50,11 +50,20 @@
         <Exec Command="&quot;$(_TransformExe)&quot; &quot;%(_TextTransform.FullPath)&quot;"
               Condition="'%(_TextTransform.Identity)' != ''"/>
 
-        <Move
-            SourceFiles="%(_TextTransform.Identity).clarius_bak"
-            DestinationFiles="%(_TextTransform.Identity)"
-        />
-
+        <CallTarget Targets="RestoreBackups" />
+        <OnError ExecuteTargets="RestoreBackups" />
     </Target>
-    
+
+    <Target Name="RestoreBackups">
+      <ItemGroup>
+        <_TextTransform
+          Include="@(None)"
+          Condition="'%(None.Generator)' == 'TextTemplatingFileGenerator'"
+        />
+      </ItemGroup>
+      <Move
+        SourceFiles="%(_TextTransform.Identity).clarius_bak"
+        DestinationFiles="%(_TextTransform.Identity)"
+      />
+    </Target>
 </Project>

--- a/nuget/build/Clarius.TransformOnBuild.targets
+++ b/nuget/build/Clarius.TransformOnBuild.targets
@@ -40,6 +40,35 @@
       Condition="!Exists('$(_TransformExe)')">$(_CommonProgramFiles)\Microsoft Shared\TextTemplating\14.0\TextTransform.exe</_TransformExe>
   </PropertyGroup>
 
+  <UsingTask
+    TaskName="SubstituteTemplateProperties"
+    TaskFactory="CodeTaskFactory"
+    AssemblyFile="$(MSBuildToolsPath)\Microsoft.Build.Tasks.v4.0.dll" >
+    <ParameterGroup>
+      <TemplateFile
+        ParameterType="System.String"
+        Required="true"
+      />
+    </ParameterGroup>
+    <Task>
+      <Using
+        Namespace="System.IO"
+      />
+      <Using
+        Namespace="System.Text.RegularExpressions"
+      />
+      <Code
+        Type="Fragment"
+        Language="cs">
+<![CDATA[
+    var template = File.ReadAllText(TemplateFile);
+    template = Regex.Replace(template, @"\$\(SolutionDir\)", "");
+    File.WriteAllText(TemplateFile, template);
+]]>
+      </Code>
+    </Task>
+  </UsingTask>
+
   <Target
     Name="TransformOnBuild"
     AfterTargets="BeforeCompile">
@@ -58,6 +87,10 @@
     <Copy
       SourceFiles="%(_TextTransform.FullPath)"
       DestinationFiles="%(_TextTransform.FullPath).clarius_bak"
+    />
+
+    <SubstituteTemplateProperties
+      TemplateFile="%(_TextTransform.FullPath)"
     />
 
     <!-- Perform task batching for each file -->

--- a/nuget/build/Clarius.TransformOnBuild.targets
+++ b/nuget/build/Clarius.TransformOnBuild.targets
@@ -41,6 +41,11 @@
                             Condition="'%(None.Generator)' == 'TextTemplatingFileGenerator'" />
         </ItemGroup>
 
+        <Copy
+            SourceFiles="%(_TextTransform.Identity)"
+            DestinationFiles="%(_TextTransform.Identity).clarius_bak"
+        />
+
         <!-- Perform task batching for each file -->
         <Exec Command="&quot;$(_TransformExe)&quot; &quot;%(_TextTransform.FullPath)&quot;"
               Condition="'%(_TextTransform.Identity)' != ''"/>

--- a/nuget/build/Clarius.TransformOnBuild.targets
+++ b/nuget/build/Clarius.TransformOnBuild.targets
@@ -35,8 +35,10 @@
   <Target
     Name="TransformOnBuild"
     AfterTargets="BeforeBuild">
-    <CallTarget
+    <MSBuild
+      Projects="$(MSBuildProjectFile)"
       Targets="ResolveReferences;_CopyFilesMarkedCopyLocal"
+      Properties="FakePropertyToEnsureThoseTargetsWillRunAgain=true"
     />
     <TransformOnBuildTask />
   </Target>

--- a/nuget/build/Clarius.TransformOnBuild.targets
+++ b/nuget/build/Clarius.TransformOnBuild.targets
@@ -41,6 +41,11 @@
   </PropertyGroup>
 
   <UsingTask
+    TaskName="Clarius.TransformOnBuild.MSBuild.Task.TransformOnBuildTask" 
+    AssemblyFile="Clarius.TransformOnBuild.MSBuild.Task.dll"
+  />
+
+  <UsingTask
     TaskName="SubstituteTemplateProperties"
     TaskFactory="CodeTaskFactory"
     AssemblyFile="$(MSBuildToolsPath)\Microsoft.Build.Tasks.v4.0.dll" >
@@ -72,6 +77,8 @@
   <Target
     Name="TransformOnBuild"
     AfterTargets="BeforeCompile">
+    <TransformOnBuildTask />
+
     <Error
       Text="Failed to find TextTransform.exe tool at '$(_TransformExe)." 
       Condition="!Exists('$(_TransformExe)')"

--- a/nuget/build/Clarius.TransformOnBuild.targets
+++ b/nuget/build/Clarius.TransformOnBuild.targets
@@ -56,8 +56,8 @@
     </ItemGroup>
 
     <Copy
-      SourceFiles="%(_TextTransform.Identity)"
-      DestinationFiles="%(_TextTransform.Identity).clarius_bak"
+      SourceFiles="%(_TextTransform.FullPath)"
+      DestinationFiles="%(_TextTransform.FullPath).clarius_bak"
     />
 
     <!-- Perform task batching for each file -->
@@ -83,8 +83,8 @@
       />
     </ItemGroup>
     <Move
-      SourceFiles="%(_TextTransform.Identity).clarius_bak"
-      DestinationFiles="%(_TextTransform.Identity)"
+      SourceFiles="%(_TextTransform.FullPath).clarius_bak"
+      DestinationFiles="%(_TextTransform.FullPath)"
     />
   </Target>
 </Project>

--- a/nuget/build/Clarius.TransformOnBuild.targets
+++ b/nuget/build/Clarius.TransformOnBuild.targets
@@ -17,114 +17,13 @@
 <Project
   ToolsVersion="4.0"
   xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
-  <PropertyGroup>
-    <_CommonProgramFiles>$([System.Environment]::GetEnvironmentVariable('CommonProgramFiles(x86)'))</_CommonProgramFiles>
-    <_CommonProgramFiles
-      Condition=" '$(_CommonProgramFiles)' == '' ">$(CommonProgramFiles)</_CommonProgramFiles>
-    <TextTransformPath
-      Condition="'$(TextTransformPath)' == ''">$(_CommonProgramFiles)\Microsoft Shared\TextTemplating\$(VisualStudioVersion)\TextTransform.exe</TextTransformPath>
-
-    <!-- Initial default value -->
-    <_TransformExe>$(TextTransformPath)</_TransformExe>
-    <!-- Cascading probing if file not found -->
-    <_TransformExe
-      Condition="!Exists('$(_TransformExe)')">$(_CommonProgramFiles)\Microsoft Shared\TextTemplating\10.0\TextTransform.exe</_TransformExe>
-    <_TransformExe
-      Condition="!Exists('$(_TransformExe)')">$(_CommonProgramFiles)\Microsoft Shared\TextTemplating\11.0\TextTransform.exe</_TransformExe>
-    <_TransformExe
-      Condition="!Exists('$(_TransformExe)')">$(_CommonProgramFiles)\Microsoft Shared\TextTemplating\12.0\TextTransform.exe</_TransformExe>
-    <!-- Future proof 'til VS2013+2 -->
-    <_TransformExe
-      Condition="!Exists('$(_TransformExe)')">$(_CommonProgramFiles)\Microsoft Shared\TextTemplating\13.0\TextTransform.exe</_TransformExe>
-    <_TransformExe
-      Condition="!Exists('$(_TransformExe)')">$(_CommonProgramFiles)\Microsoft Shared\TextTemplating\14.0\TextTransform.exe</_TransformExe>
-  </PropertyGroup>
-
   <UsingTask
     TaskName="Clarius.TransformOnBuild.MSBuild.Task.TransformOnBuildTask" 
     AssemblyFile="Clarius.TransformOnBuild.MSBuild.Task.dll"
   />
-
-  <UsingTask
-    TaskName="SubstituteTemplateProperties"
-    TaskFactory="CodeTaskFactory"
-    AssemblyFile="$(MSBuildToolsPath)\Microsoft.Build.Tasks.v4.0.dll" >
-    <ParameterGroup>
-      <TemplateFile
-        ParameterType="System.String"
-        Required="true"
-      />
-    </ParameterGroup>
-    <Task>
-      <Using
-        Namespace="System.IO"
-      />
-      <Using
-        Namespace="System.Text.RegularExpressions"
-      />
-      <Code
-        Type="Fragment"
-        Language="cs">
-<![CDATA[
-    var template = File.ReadAllText(TemplateFile);
-    template = Regex.Replace(template, @"\$\(SolutionDir\)", "");
-    File.WriteAllText(TemplateFile, template);
-]]>
-      </Code>
-    </Task>
-  </UsingTask>
-
   <Target
     Name="TransformOnBuild"
     AfterTargets="BeforeCompile">
     <TransformOnBuildTask />
-
-    <Error
-      Text="Failed to find TextTransform.exe tool at '$(_TransformExe)." 
-      Condition="!Exists('$(_TransformExe)')"
-    />
-
-    <ItemGroup>
-      <_TextTransform
-        Include="@(None)"
-        Condition="'%(None.Generator)' == 'TextTemplatingFileGenerator'"
-      />
-    </ItemGroup>
-
-    <Copy
-      SourceFiles="%(_TextTransform.FullPath)"
-      DestinationFiles="%(_TextTransform.FullPath).clarius_bak"
-    />
-
-    <SubstituteTemplateProperties
-      TemplateFile="%(_TextTransform.FullPath)"
-    />
-
-    <!-- Perform task batching for each file -->
-    <Exec
-      Command="&quot;$(_TransformExe)&quot; &quot;%(_TextTransform.FullPath)&quot;"
-      Condition="'%(_TextTransform.Identity)' != ''"
-    />
-
-    <CallTarget
-      Targets="RestoreBackups"
-    />
-
-    <OnError
-      ExecuteTargets="RestoreBackups"
-    />
-  </Target>
-
-  <Target Name="RestoreBackups">
-    <ItemGroup>
-      <_TextTransform
-        Include="@(None)"
-        Condition="'%(None.Generator)' == 'TextTemplatingFileGenerator'"
-      />
-    </ItemGroup>
-    <Move
-      SourceFiles="%(_TextTransform.FullPath).clarius_bak"
-      DestinationFiles="%(_TextTransform.FullPath)"
-    />
   </Target>
 </Project>

--- a/nuget/build/Clarius.TransformOnBuild.targets
+++ b/nuget/build/Clarius.TransformOnBuild.targets
@@ -34,9 +34,9 @@
   />
   <Target
     Name="TransformOnBuild"
-    AfterTargets="BeforeCompile">
+    AfterTargets="BeforeBuild">
     <CallTarget
-      Targets="_CopyFilesMarkedCopyLocal"
+      Targets="ResolveReferences;_CopyFilesMarkedCopyLocal"
     />
     <TransformOnBuildTask />
   </Target>

--- a/nuget/build/Clarius.TransformOnBuild.targets
+++ b/nuget/build/Clarius.TransformOnBuild.targets
@@ -14,56 +14,77 @@
    See the License for the specific language governing permissions and
    limitations under the License.
 -->
-<Project ToolsVersion="4.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
-	
-	<PropertyGroup>
-        <_CommonProgramFiles>$([System.Environment]::GetEnvironmentVariable('CommonProgramFiles(x86)'))</_CommonProgramFiles>
-		<_CommonProgramFiles Condition=" '$(_CommonProgramFiles)' == '' ">$(CommonProgramFiles)</_CommonProgramFiles>		
-		<TextTransformPath Condition="'$(TextTransformPath)' == ''">$(_CommonProgramFiles)\Microsoft Shared\TextTemplating\$(VisualStudioVersion)\TextTransform.exe</TextTransformPath>
-        <!-- Initial default value -->
-		<_TransformExe>$(TextTransformPath)</_TransformExe>
-        <!-- Cascading probing if file not found -->
-        <_TransformExe Condition="!Exists('$(_TransformExe)')">$(_CommonProgramFiles)\Microsoft Shared\TextTemplating\10.0\TextTransform.exe</_TransformExe>
-        <_TransformExe Condition="!Exists('$(_TransformExe)')">$(_CommonProgramFiles)\Microsoft Shared\TextTemplating\11.0\TextTransform.exe</_TransformExe>
-        <_TransformExe Condition="!Exists('$(_TransformExe)')">$(_CommonProgramFiles)\Microsoft Shared\TextTemplating\12.0\TextTransform.exe</_TransformExe>
-        <!-- Future proof 'til VS2013+2 -->
-        <_TransformExe Condition="!Exists('$(_TransformExe)')">$(_CommonProgramFiles)\Microsoft Shared\TextTemplating\13.0\TextTransform.exe</_TransformExe>
-        <_TransformExe Condition="!Exists('$(_TransformExe)')">$(_CommonProgramFiles)\Microsoft Shared\TextTemplating\14.0\TextTransform.exe</_TransformExe>
-	</PropertyGroup>
+<Project
+  ToolsVersion="4.0"
+  xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <PropertyGroup>
+    <_CommonProgramFiles>$([System.Environment]::GetEnvironmentVariable('CommonProgramFiles(x86)'))</_CommonProgramFiles>
+    <_CommonProgramFiles
+      Condition=" '$(_CommonProgramFiles)' == '' ">$(CommonProgramFiles)</_CommonProgramFiles>
+    <TextTransformPath
+      Condition="'$(TextTransformPath)' == ''">$(_CommonProgramFiles)\Microsoft Shared\TextTemplating\$(VisualStudioVersion)\TextTransform.exe</TextTransformPath>
 
-    <Target Name="TransformOnBuild" AfterTargets="BeforeCompile">
+    <!-- Initial default value -->
+    <_TransformExe>$(TextTransformPath)</_TransformExe>
+    <!-- Cascading probing if file not found -->
+    <_TransformExe
+      Condition="!Exists('$(_TransformExe)')">$(_CommonProgramFiles)\Microsoft Shared\TextTemplating\10.0\TextTransform.exe</_TransformExe>
+    <_TransformExe
+      Condition="!Exists('$(_TransformExe)')">$(_CommonProgramFiles)\Microsoft Shared\TextTemplating\11.0\TextTransform.exe</_TransformExe>
+    <_TransformExe
+      Condition="!Exists('$(_TransformExe)')">$(_CommonProgramFiles)\Microsoft Shared\TextTemplating\12.0\TextTransform.exe</_TransformExe>
+    <!-- Future proof 'til VS2013+2 -->
+    <_TransformExe
+      Condition="!Exists('$(_TransformExe)')">$(_CommonProgramFiles)\Microsoft Shared\TextTemplating\13.0\TextTransform.exe</_TransformExe>
+    <_TransformExe
+      Condition="!Exists('$(_TransformExe)')">$(_CommonProgramFiles)\Microsoft Shared\TextTemplating\14.0\TextTransform.exe</_TransformExe>
+  </PropertyGroup>
 
-        <Error Text="Failed to find TextTransform.exe tool at '$(_TransformExe)." 
-                Condition="!Exists('$(_TransformExe)')"/>
+  <Target
+    Name="TransformOnBuild"
+    AfterTargets="BeforeCompile">
+    <Error
+      Text="Failed to find TextTransform.exe tool at '$(_TransformExe)." 
+      Condition="!Exists('$(_TransformExe)')"
+    />
 
-        <ItemGroup>
-            <_TextTransform Include="@(None)"
-                            Condition="'%(None.Generator)' == 'TextTemplatingFileGenerator'" />
-        </ItemGroup>
-
-        <Copy
-            SourceFiles="%(_TextTransform.Identity)"
-            DestinationFiles="%(_TextTransform.Identity).clarius_bak"
-        />
-
-        <!-- Perform task batching for each file -->
-        <Exec Command="&quot;$(_TransformExe)&quot; &quot;%(_TextTransform.FullPath)&quot;"
-              Condition="'%(_TextTransform.Identity)' != ''"/>
-
-        <CallTarget Targets="RestoreBackups" />
-        <OnError ExecuteTargets="RestoreBackups" />
-    </Target>
-
-    <Target Name="RestoreBackups">
-      <ItemGroup>
-        <_TextTransform
-          Include="@(None)"
-          Condition="'%(None.Generator)' == 'TextTemplatingFileGenerator'"
-        />
-      </ItemGroup>
-      <Move
-        SourceFiles="%(_TextTransform.Identity).clarius_bak"
-        DestinationFiles="%(_TextTransform.Identity)"
+    <ItemGroup>
+      <_TextTransform
+        Include="@(None)"
+        Condition="'%(None.Generator)' == 'TextTemplatingFileGenerator'"
       />
-    </Target>
+    </ItemGroup>
+
+    <Copy
+      SourceFiles="%(_TextTransform.Identity)"
+      DestinationFiles="%(_TextTransform.Identity).clarius_bak"
+    />
+
+    <!-- Perform task batching for each file -->
+    <Exec
+      Command="&quot;$(_TransformExe)&quot; &quot;%(_TextTransform.FullPath)&quot;"
+      Condition="'%(_TextTransform.Identity)' != ''"
+    />
+
+    <CallTarget
+      Targets="RestoreBackups"
+    />
+
+    <OnError
+      ExecuteTargets="RestoreBackups"
+    />
+  </Target>
+
+  <Target Name="RestoreBackups">
+    <ItemGroup>
+      <_TextTransform
+        Include="@(None)"
+        Condition="'%(None.Generator)' == 'TextTemplatingFileGenerator'"
+      />
+    </ItemGroup>
+    <Move
+      SourceFiles="%(_TextTransform.Identity).clarius_bak"
+      DestinationFiles="%(_TextTransform.Identity)"
+    />
+  </Target>
 </Project>

--- a/nuget/build/Clarius.TransformOnBuild.targets
+++ b/nuget/build/Clarius.TransformOnBuild.targets
@@ -49,7 +49,12 @@
         <!-- Perform task batching for each file -->
         <Exec Command="&quot;$(_TransformExe)&quot; &quot;%(_TextTransform.FullPath)&quot;"
               Condition="'%(_TextTransform.Identity)' != ''"/>
-        
+
+        <Move
+            SourceFiles="%(_TextTransform.Identity).clarius_bak"
+            DestinationFiles="%(_TextTransform.Identity)"
+        />
+
     </Target>
     
 </Project>

--- a/nuget/build/Clarius.TransformOnBuild.targets
+++ b/nuget/build/Clarius.TransformOnBuild.targets
@@ -17,9 +17,20 @@
 <Project
   ToolsVersion="4.0"
   xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <PropertyGroup>
+    <TempFolder>$([System.IO.Path]::GetTempPath())$([System.Guid]::NewGuid())</TempFolder>
+  </PropertyGroup>
+  <Target
+    Name="CopyTaskAssemblyToTempFolder"
+    BeforeTargets="BeforeBuild">
+    <Copy
+      SourceFiles="$(MSBuildThisFileDirectory)Clarius.TransformOnBuild.MSBuild.Task.dll"
+      DestinationFolder="$(TempFolder)"
+    />
+  </Target>
   <UsingTask
     TaskName="Clarius.TransformOnBuild.MSBuild.Task.TransformOnBuildTask" 
-    AssemblyFile="Clarius.TransformOnBuild.MSBuild.Task.dll"
+    AssemblyFile="$(TempFolder)\Clarius.TransformOnBuild.MSBuild.Task.dll"
   />
   <Target
     Name="TransformOnBuild"

